lib/blob/blobstore.c

```diff
@@ -7202,6 +7259,232 @@ spdk_bs_blob_shallow_copy(struct spdk_blob_store *bs, struct spdk_io_channel *ch
+		ctx->bserrno = bserrno;
+	}
+
+	bs_set_parent_cleanup_finish(ctx, 0);
+}
+
+static void
+bs_set_parent_close_blob(void *cb_arg, int bserrno)
+{
+	struct set_parent_ctx *ctx = cb_arg;
+	struct spdk_blob *blob = ctx->blob;
+	struct spdk_blob *snapshot = ctx->parent.blob;
+
+	if (bserrno != 0 && ctx->bserrno == 0) {
+		SPDK_ERRLOG("error %d in metadata sync\n", bserrno);
+		ctx->bserrno = bserrno;
+	}
+
+	blob->locked_operation_in_progress = false;
+	snapshot->locked_operation_in_progress = false;
+
+	spdk_blob_close(blob, bs_set_parent_close_snapshot, ctx);
+}
+
+static void
+bs_set_parent_unfrozen(void *cb_arg, int bserrno)
+{
+	struct set_parent_ctx *ctx = cb_arg;
+	struct spdk_blob *blob = ctx->blob;
+
+	if (bserrno != 0) {
+		SPDK_ERRLOG("error %d setting back_bs_dev\n", bserrno);
+		ctx->bserrno = bserrno;
+		bs_set_parent_close_blob(ctx, bserrno);
+		return;
+	}
+
+	spdk_blob_sync_md(blob, bs_set_parent_close_blob, ctx);
+}
+
+static int
+bs_set_parent_refs(struct spdk_blob *blob, struct blob_parent_id *parent_id)
+{
+	int rc;
+
+	bs_blob_list_remove(blob);
+
+	if (blob_is_esnap_clone(blob)) {
+		/* Remove the xattr that references the external snapshot */
+		blob->invalid_flags &= ~SPDK_BLOB_EXTERNAL_SNAPSHOT;
+		blob_remove_xattr(blob, BLOB_EXTERNAL_SNAPSHOT_ID, true);
+	}
+
+	rc = blob_set_xattr(blob, BLOB_SNAPSHOT, &parent_id->u.snapshot.id, sizeof(spdk_blob_id), true);
+	if (rc != 0) {
+		SPDK_ERRLOG("error %d setting snapshot xattr\n", rc);
+		free(parent_id);
+		return rc;
+	}
+
+	blob->parent_id = parent_id->u.snapshot.id;
+
+	spdk_blob_set_read_only(blob);
+
+	bs_blob_list_add(blob);
+
+	free(parent_id);
+	return 0;
+}
+
+static void
+bs_set_parent_snapshot_open_cpl(void *cb_arg, struct spdk_blob *snapshot, int bserrno)
+{
+	struct set_parent_ctx *ctx = cb_arg;
+	struct spdk_blob *blob = ctx->blob;
+	struct spdk_bs_dev *back_bs_dev;
+	struct blob_parent_id *parent_id;
+
+	if (bserrno != 0) {
+		SPDK_ERRLOG("snapshot open error %d\n", bserrno);
+		ctx->bserrno = bserrno;
+		spdk_blob_close(blob, bs_set_parent_cleanup_finish, ctx);
+		return;
+	}
+
+	ctx->parent.blob = snapshot;
+
+	if (!spdk_blob_is_snapshot(snapshot)) {
+		SPDK_ERRLOG("parent blob is not a snapshot\n");
+		ctx->bserrno = -EINVAL;
+		spdk_blob_close(blob, bs_set_parent_close_snapshot, ctx);
+		return;
+	}
+
+	if (blob->active.num_clusters != snapshot->active.num_clusters) {
+		SPDK_ERRLOG("parent blob has a number of clusters different from child's ones\n");
+		ctx->bserrno = -EINVAL;
+		spdk_blob_close(blob, bs_set_parent_close_snapshot, ctx);
+		return;
+	}
+
+	if (blob->locked_operation_in_progress || snapshot->locked_operation_in_progress) {
+		SPDK_ERRLOG("cannot set local parent of blob, another operation in progress\n");
+		ctx->bserrno = -EBUSY;
+		spdk_blob_close(blob, bs_set_parent_close_snapshot, ctx);
+		return;
+	}
+
+	blob->locked_operation_in_progress = true;
+	snapshot->locked_operation_in_progress = true;
```

Comment:
No, because metadata operations are executed in the same spdk_thread

Reply:
Then one minor question: is it better to move the blob check and locking into func `bs_set_parent_blob_open_cpl`?

Reply:
No, mainly because to set the lock we need the snapshot opened. Moreover we must perform some checks over the snapshot that could abort the operation and it is better to do that before to acquire the lock.